### PR TITLE
Adding Branch Labels Auto-associates Project Board

### PR DESCRIPTION
**Pull request by Hanaffi**
Closes https://github.com/electron/cation/issues/91.

Checklist: 
- [x] Find associated projects names
- [x] Handle case when project isn't created yet
- [x] Handle case when Unsorted column isn't created yet
- [x] Handle case when issue gets closed

### CODE REVIEW

File: src/add-card-to-project.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            const repo = payload.repository.name;
+            const owner = payload.sender.login;
+            const issue_number = payload.issue.number;
+
+            const { data: projects } = await octokit.projects.listForRepo({ owner, repo, });
+
+            const project_obj = projects.find(
+                // WARNING: This is just for trial.
+                // TODO: find a way to get issue is associated tags
+                (el:any) => el.name == "x-18-19"
```
Comment (codebytere) [inline src/add-card-to-project.ts:18]: branch names are always `number-x-y` - something like `x-18-19` would never exist. Examples:

- 17-x-y
- 18-x-y
- 19-x-y

File: src/add-card-to-project.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { Probot,Context } from 'probot';
+import { LogLevel } from './enums';
+import { log } from './utils/log-util';
+
+export function addCardToProject(probot: Probot) {
+    probot.on("issues.opened", async (context:Context['octokit']) => {
```
Comment (codebytere) [inline src/add-card-to-project.ts:6]: The association would occur at the point the issue is labeled, not opened - this should be `issues.opened`.

Hooking on this event also gives us access to the specific label that was added - this then tells us which board we would be associating it to. You can pull `label` from `context.payload` to get that information.

File: src/add-card-to-project.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            
+            await octokit.projects.createCard({
+                column_id: unsorted_column.id, // This is the id of the Unsorted Issues column, which you'll need to get
+                content_id: issue_number, // This is the issue number
+                content_type: "Issue", // This will always be issue since we're associating an issue
+            });
+           
+        } catch (err: any) {
+            // TODO: Change error message to something meaningful
+            log('addCardToProject', LogLevel.ERROR, `Something went wrong..`);
```
Comment (codebytere) [inline src/add-card-to-project.ts:46]: Something like:

```
log('addCardToProject', LogLevel.ERROR, 'Unable to associate issue with project board: ', err);
```

is fine.

File: src/add-card-to-project.ts (hunk 1 above)
Comment (codebytere) [inline src/add-card-to-project.ts:18]: Let's type this more specifically with Octokit

File: src/add-card-to-project.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            const { data: projects } = await octokit.projects.listForRepo({ owner, repo, });
+
+            const project_obj = projects.find(
+                // WARNING: This is just for trial.
+                // TODO: find a way to get issue is associated tags
+                (el:any) => el.name == "x-18-19"
+            );
+                    
+            if (!project_obj) {
+                // TODO: Create new project if it doesn't exist
```
Comment (codebytere) [inline src/add-card-to-project.ts:22]: We probably wouldn't want to create a new project if it doesn't exist. Logging and bailing is fine:

```
log('addCardToProject', LogLevel.ERROR, `Unable to find project board with name ${name}`);
```
where `name` is the label name that was added to the issue.

File: src/add-card-to-project.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                throw new Error("No project found!");
+            }
+
+            const project_id = project_obj.id;
+
+            const { data: project_columns } = await octokit.projects.listColumns({ project_id,});
+
+            const unsorted_column = project_columns.find( (el:any) => el.name == "unsorted" );
```
Comment (codebytere) [inline src/add-card-to-project.ts:30]: s/unsorted/Unsorted

**Review by codebytere — CHANGES_REQUESTED** [review]
Added some comments to point you in the right direction!

**Comment by Hanaffi** [comment]
Hey @codebytere , Can you please check it if you have time? :)

**Comment by Hanaffi** [comment]
@MarshallOfSound Probably Shelley won't be available. Can you please review this when you have time?

**Comment by Hanaffi** [comment]
@codebytere Ping

File: src/constants.ts [hunk 6]
```diff
@@ -54,3 +54,7 @@ export const API_WORKING_GROUP = 'wg-api';
 export const EXCLUDE_LABELS = [BACKPORT_LABEL, BACKPORT_SKIP_LABEL, FAST_TRACK_LABEL];
 export const EXCLUDE_PREFIXES = ['build', 'ci', 'test'];
 export const EXCLUDE_USERS = ['roller-bot[bot]', 'electron-bot'];
+
+// REGEX
+export const PROJECT_NAME_REGEX = /^[1-9]+[0-9]*-x-y$/;
```
Comment (codebytere) [inline src/constants.ts:59]: This should be something like `BRANCH_NAME_REGEX` since that's more accurate.

File: src/add-card-to-project.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+export function addCardToProject(probot: Probot) {
+  probot.on('issues.labeled', async context => {
+    try {
+      const { payload, octokit } = context;
+      const { name: repo } = payload.repository;
+      const { login: owner } = payload.sender;
+      const { number: issueNumber } = payload.issue;
+      const { name: labelName } = payload.label || {};
+
+      if (labelName == null) {
```
Comment (codebytere) [inline src/add-card-to-project.ts:16]: We don't need both of these conditionals, and we probably don't want to throw a bespoke error here.

```js
if (!label || !PROJECT_NAME_REGEX.test(label.name)) {
   log('addCardToProject', LogLevel.ERROR, `No project with ${label?.name} exists`);
   return;
}
```

should be sufficient.

File: src/add-card-to-project.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        log('addCardToProject', LogLevel.ERROR, `No label found`);
+      }
+
+      if (!PROJECT_NAME_REGEX.test(labelName!)) {
+        throw new Error(`Label doesn't match a project label!`);
+      }
+
+      const { data: projects } = await octokit.projects.listForRepo({ owner, repo });
```
Comment (codebytere) [inline src/add-card-to-project.ts:24]: We can use:
```js
const { data: projects } = await octokit.projects.listForRepo(context.repo({}));
```

here.

File: src/add-card-to-project.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        (project: EventPayloads.WebhookPayloadProjectProject) => project.name === labelName,
+      );
+
+      if (!project) {
+        log(
+          'addCardToProject',
+          LogLevel.ERROR,
+          `Unable to find project board with name ${labelName}`,
+        );
+        throw new Error(`Unable to find project board with name ${labelName}`);
```
Comment (codebytere) [inline src/add-card-to-project.ts:36]: we can simply return here as well.

```js
if (!project) {
  log(
    'addCardToProject',
    LogLevel.ERROR,
    `Unable to find project board with name ${label.name}`,
  );
  return;
}
```

File: src/add-card-to-project.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        project_id: project.id,
+      });
+
+      const unsortedColumn = projectColumns.find(
+        (column: EventPayloads.WebhookPayloadProjectColumnProjectColumn) =>
+          UNSORTED_COLUMN_ALT_NAME_REGEX.test(column.name),
+      );
+
+      if (!unsortedColumn) {
+        log('addCardToProject', LogLevel.ERROR, `Unsorted column doesn't exist!`);
```
Comment (codebytere) [inline src/add-card-to-project.ts:49]: Here as well.

```js
      if (!unsortedColumn) {
        log('addCardToProject', LogLevel.ERROR, `Unsorted column doesn't exist!`);
        return;
      }
```

File: src/add-card-to-project.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        throw new Error("Unsorted column doesn't exist");
+      }
+
+      await octokit.projects.createCard({
+        column_id: unsortedColumn.id, // This is the id of the Unsorted Issues column, which you'll need to get
+        content_id: issueNumber, // This is the issue number
```
Comment (codebytere) [inline src/add-card-to-project.ts:55]: These comments can be removed.

**Comment by codebytere** [comment]
Please also rebase on main and run `npm run prettier:write` on your branch to lint it properly.

**Comment by codebytere** [comment]
Closing as we have decided to go a different direction with this feature. We appreciate your efforts here, however, as well as your willingness to work with our change requests!

**Pull request closed without merging**